airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/state_converter.py

```diff
@@ -117,27 +212,44 @@ def convert_from_sequential_state(self, stream_state: MutableMapping[str, Any])
             "legacy": stream_state,
         }
 
-    def convert_to_sequential_state(self, stream_state: MutableMapping[str, Any]) -> Any:
+    def convert_to_sequential_state(self, stream_state: MutableMapping[str, Any]) -> MutableMapping[str, Any]:
         """
+        Convert the state message from the concurrency-compatible format to the stream's original format.
+
         e.g.
-        {
-            "state_type": "date-range",
-            "metadata": { … },
-            "slices": [
-                {starts: 0, end: 1617030403, finished_processing: true}
-            ]
-        }
-        =>
-        { "created": 1617030403 }
+        { "created": "2021-01-18T21:18:20.000Z" }
         """
         if self.is_state_message_compatible(stream_state):
             legacy_state = stream_state.get("legacy", {})
             if slices := stream_state.pop("slices", None):
-                legacy_state.update({self._cursor_field: self._get_latest_complete_time(slices)})
-            return legacy_state
+                legacy_state.update({self._cursor_field: self._millis_format(self._get_latest_complete_time(slices))})
+            return legacy_state or {}
         else:
             return stream_state
 
-    @staticmethod
-    def increment(timestamp: Any) -> Any:
-        return timestamp + 1
+    def _millis_format(self, timestamp: Optional[str]) -> Optional[str]:
+        return pendulum.parse(timestamp).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z" if timestamp else None
+
+    def _get_latest_complete_time(self, slices: List[MutableMapping[str, Any]]) -> Optional[str]:
```

Comment:
the only difference between this implementation and the `EpochValueConcurrentStreamStateConverter`'s is the return type, right?

I wonder if we could decouple the datetimes from how they're formatted since the two classes essentially have the same logic and only differ in the 
- output format
- the zero value (unix epoc vs datetime(1, 1, 1, 0, 0, 0, 0)) 
- the size of the increment (1 second vs 1ms)

I don't think it would be a significant change, we'd just need to keep the slice boundaries as datetimes and format them in the convert methods

Reply:
Removing an earlier comment to share it here because it feels inline with what Alex is mentioning:

It feels like `EpochValueConcurrentStreamStateConverter.convert_from_sequential_state` is somewhat identical to `IsoMillisConcurrentStreamStateConverter.convert_from_sequential_state`. The same is true for `convert_to_sequential_state`. It makes sense to me as they are both converters for datetime. Would it mean that they should all share the same logic and the only difference would be how we adapt/assemble this to an AirbyteStateMessage? My concern is that if one change, the other will most probably change as well